### ISSUE TITLE
Realtek ASIO Issues

### ISSUE THREAD

**Issue by AKPWebDesign** [eiz/SynchronousAudioRouter#14]
So, I've set up SAR on my system with pretty great success, using ASIO4ALL. The only problem is that no matter how low I set the buffer size, there's always enough of a delay that I can't properly speak when I've got my microphone playing back into my headphones. To try to solve the issue, I decided to try using the Realtek ASIO driver that my computer has available.

This works well within REAPER. I am able to have my mic monitoring on and can't hear even the slightest bit of a delay. The problem is when I attempt to have any sounds come through the interfaces created by SAR. I've recorded what it sounds like when I send the Windows audio test through, available [here](https://drive.google.com/file/d/0B5wWwqnenRLpRkJkQ0xvMFM4SVk/view?usp=sharing).

Any ideas what this could be caused by, and/or how it can be fixed? Thanks!

**Comment by eiz** [comment]
What audio hardware are you using?

**Comment by AKPWebDesign** [comment]
Just the built-in headphone/microphone combo jack, no external hardware.

**Comment by eiz** [comment]
Yes. What motherboard, drivers, etc? The only references I see to Realtek having an ASIO driver are some unmaintained thing from 2007. If you go in Device Manager and find your sound device and open its properties, what IDs do you see on the Details tab for the "Hardware Ids" property?

**Comment by AKPWebDesign** [comment]
Ah, yes. Sorry about that, just got off work and my brain's a little fried. To start, it's a Dell XPS 15 9560 laptop, not sure if I can find mobo information from Dell about it. I've tried a variety of drivers, but currently using the latest Realtek drivers from Dell for my laptop. I've also tried using the latest from Realtek directly, but the same problem occurred.

As for Hardware IDs, I've got a single 'Realtek Audio' device with IDs `INTELAUDIO\FUNC_01&VEN_10EC&DEV_0298&SUBSYS_102807BE&REV_1001` and `INTELAUDIO\FUNC_01&VEN_10EC&DEV_0298&SUBSYS_102807BE`

**Comment by eiz** [comment]
This is most likely related to a silly assumption SarAsio makes about the sample types supported by the underlying device. There's a TODO in the code that needs to be fixed. I have a similar Realtek HD Audio device that reproduces a similar problem (different distortion, but still sounds very bad).
